Replace Flipflop with Sail

vinistock opened the pull request:
<!--- Prepend PR title with [WIP] if work in progress. Remove when ready for review. -->

## What type of PR is this? (check all applicable)

- [X] Refactor
- [X] Feature
- [ ] Bug Fix
- [ ] Documentation Update

## Description

Replaces Flipflop with Sail as described in the related issue. The idea is to allow people to take a look at Sail and see if it fully fits the needs.

I am available to make changes to Sail if need be and also expand the work here.

This is my first contribution to dev.to, so please let me know if I'm breaking any code conventions.

## Related Tickets & Documents

Resolves #686 

## Added to documentation?

- [ ] docs.dev.to
- [ ] readme
- [x] no documentation needed

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/thepracticaldev/dev.to?pullRequest=841) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/thepracticaldev/dev.to?pullRequest=841) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/thepracticaldev/dev.to?pullRequest=841) it.</sub>

benhalpern commented:
@jessleenyc wanna give this one a look? More context in the linked issue.

jessleenyc commented on config/application.rb:
```diff
@@ -40,12 +40,6 @@ class Application < Rails::Application
-    config.flipflop.dashboard_access_filter = -> {
-      head :forbidden unless current_user.has_any_role?(:super_admin)
-    }
-
```
This made it so that only super admins had access to toggle feature flags. Is there something similar we can add for sail?

vinistock replied:
I will add the support for defining authorization lambdas in Sail

jessleenyc replied:
Awesome

vinistock commented on db/migrate/20181001225906_create_sail_settings.rb:
```diff
@@ -0,0 +1,22 @@
+        { name: :upcoming_events, cast_type: :boolean, value: true, description: "Toggle upcoming events in sidebar" }
+    ].each { |setting_info| Sail::Setting.create(setting_info) }
+
+    drop_table :flipflop_features
```
Do you want simply a different migration file or would you rather drop the table in a distinct commit?

jessleenyc replied:
Yep, a standalone migration file for dropping the flip flop table would be great. Doesn't need to be a distinct commit.

vinistock commented:
@jessleenyc you should have seen the settings when you visited the dashboard. Did you database get populated properly after you ran the migration? I will double check the migration to see if it is populating the database.

jessleenyc commented:
@vinistock I ran the migration and it looks like it's in the db properly. Were there more steps I may have missed?

![](https://cl.ly/f9e7fd299df7/download/Image%202018-10-10%20at%204.35.42%20PM.png)

vinistock commented:
@jessleenyc I have pushed a commit addressing your concerns. Let me know what you think and if there's anything else you wish to change. However, it seems the build and the codeclimate report might have gotten stuck for some reason.

maestromac commented on config/routes.rb:
```diff
@@ -20,6 +20,7 @@
 
     root controller: DashboardManifest::ROOT_DASHBOARD, action: :index
   end
+  mount Sail::Engine => "/sail"
```
@benhalpern is this endpoint alright would you like prefer it elsewhere?

maestromac commented on db/migrate/20181001225906_create_sail_settings.rb:
```diff
@@ -0,0 +1,20 @@
+        { name: :live_is_live, cast_type: :boolean, value: 'false', description: "/live page showing live event" },
+        { name: :she_coded, cast_type: :boolean, value: 'false', description: "Toggle #shecoded sidebar" },
+        { name: :upcoming_events, cast_type: :boolean, value: 'true', description: "Toggle upcoming events in sidebar" }
+    ].each { |setting_info| Sail::Setting.create!(setting_info) }
```
If, instead, we choose to create `config/sail.yml` to store these settings. Would our database automatically remove settings that were removed from `config/sail.yml`?

vinistock replied:
Yeah, sure, no problem.

vinistock commented:
@maestromac Just pushed another commit moving the setting creation from the migration to the config YAML file. Could you please do a quick re-test and make sure everything is fine?

I'd do it myself, but my development environment needs some attention for dev.to. It's complaining about missing database.yml and I see it is in the gitignore file, so my environment must be outdated.

benhalpern commented:
We hit an error while trying to merge this, so we're going to try and get Sail in one step at a time.

vinistock commented:
@benhalpern can you disclose which error happened? Is it something that could be fixed in Sail?

maestromac commented:
@vinistock we are still uncertain on what crashed production and it could be our end. We are setting up a production test environment and will deploy more Sail changes there first. I'll keep you posted on connect.
